### PR TITLE
[backend-comparison] Add system information to benchmark results

### PR DESCRIPTION
**Pull request by syl20bnr**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

https://github.com/tracel-ai/burn/issues/1072

### Changes

Add system information using `sysinfo` crate for the CPUs and `wgpu` crate for the GPUs.

### CODE REVIEW

File: backend-comparison/src/persistence/system_info.rs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    fn enumerate_gpus() -> Vec<String> {
+        let instance = wgpu::Instance::default();
+        let adapters: Vec<wgpu::Adapter> = instance
+            .enumerate_adapters(wgpu::Backends::all())
+            .filter(|adapter| {
+                let info = adapter.get_info();
+                info.device_type == wgpu::DeviceType::DiscreteGpu
+            })
```
Comment (nathanielsimard) [inline backend-comparison/src/persistence/system_info.rs:39]: I think we want to keep the integrated GPU as well. For instance, newer Macs with M1/2/3 chips use integrated GPUs (powerful ones) with shared memory.

File: backend-comparison/src/persistence/system_info.rs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            .iter()
+            .map(|c| c.brand().to_string())
+            .collect();
+        cpu_names.into_iter().collect()
+    }
+
+    fn enumerate_gpus() -> Vec<String> {
+        let instance = wgpu::Instance::default();
+        let adapters: Vec<wgpu::Adapter> = instance
+            .enumerate_adapters(wgpu::Backends::all())
```
Comment (nathanielsimard) [inline backend-comparison/src/persistence/system_info.rs:35]: Is there any duplicates when enumerating all adapters for all backends? Should we select the backend based on the OS here?

**Comment by syl20bnr** [comment]
I added the integrated GPUs and also the selection algo from burn-wgpu `AutoGraphicsApi`.

On my intel macbook I get this systeminfo:

```
  "systemInfo": {
    "cpus": [
      "Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz"
    ],
    "gpus": [
      "AMD Radeon Pro 5300M",
      "Intel(R) UHD Graphics 630"
    ]
  },

```

**Comment by syl20bnr** [comment]
We have a regression when using the `burn-wgpu` crate though, the fusion backend is automatically selected again.

**Comment by syl20bnr** [comment]
> We have a regression when using the burn-wgpu crate though, the fusion backend is automatically selected again.

Fixed by disabling the default features for `burn-wgpu` in the `backend-comparison` crate.

**Comment by syl20bnr** [comment]
Well we end up in the same situation than: https://github.com/tracel-ai/burn/pull/1476

@nathanielsimard so the solution is to remove the fusion default in wgpu ?

File: backend-comparison/src/persistence/system_info.rs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    gpus: Vec<String>,
+}
+
+impl BenchmarkSystemInfo {
+    pub(crate) fn new() -> Self {
+        Self {
+            cpus: BenchmarkSystemInfo::enumerate_cpus(),
+            gpus: BenchmarkSystemInfo::enumerate_gpus(),
+        }
+    }
```
Comment (nathanielsimard) [inline backend-comparison/src/persistence/system_info.rs:19]: Clean 👌

**Review by nathanielsimard — APPROVED** [review]

**Comment by syl20bnr** [comment]
I merged it but we still have the regression that in the current feature config we cannot enable wgpu alone in the benchmarks. It is not trivial to fix this as mentioned in https://github.com/tracel-ai/burn/pull/1476

**Pull request merged**